contrib/mod_sftp/channel.c

```diff
@@ -121,7 +123,26 @@ static struct ssh2_channel *alloc_channel(const char *type,
     channel_list = make_array(channel_pool, 1, sizeof(struct ssh2_channel *));
   }
 
-  *((struct ssh2_channel **) push_array(channel_list)) = chan;
+  /* Look for an empty slot in the list, from an already-destroyed channel,
+   * first.
+   */
+  chans = channel_list->elts;
+  for (i = 0; i < channel_list->nelts; i++) {
+    if (chans[i] == NULL) {
```

Comment:
By reusing an existing (and empty) slot in the channel list, we avoid expanding the list on every new channel.  The [`SFTPMaxChannels`](http://www.proftpd.org/docs/contrib/mod_sftp.html#SFTPMaxChannels) limit places an upper bound on the number of concurrent items in this list, so as long as we reusing empty slots, there is no excess memory allocation across thousands of opened/closed channels.